[TEP0138] Decouple v1beta1 beta feature validation

JeromeJu opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes
This PR contains 2 commits. The first decouples the existing beta 
feature validation in v1beta1. Prior to this change, beta features are 
regarded as stable in v1beta1 apiVersion and they are validated only in
 v1 api behind beta enable-api-fields but not in v1beta1. This PR removes 
the gapbetween the validations of the stable features in v1 and v1beta1.

The second commit moved Task.Validate and Pipeline.Validate to
TaskSpec.Validate and PipelineSpec.Validate.

It also updates the api compatibility policy to clarify that feature
stability levels are independent on CRD apiVersions.

Fixes: https://github.com/tektoncd/pipeline/issues/6592
Part of: #7177 

<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
v1beta1 CRDs with beta features now requires `beta` `enable-api-fields`. More specifically, users who have been accidentally using beta features resolvers, object array params and results with `enable-api-fields` set to `stable` now needs to change to `enable-api-fields=beta`
```

JeromeJu commented:
/remove-kind cleanup
/kind misc

JeromeJu commented:
/hold
wait till https://github.com/tektoncd/community/pull/1034 gets merged first

JeromeJu commented:
/assign @lbernick @chitrangpatel

chitrangpatel commented on pkg/apis/pipeline/v1beta1/pipeline_types_test.go:
```diff
@@ -223,7 +223,7 @@ func TestPipelineTask_ValidateRegularTask_Success(t *testing.T) {
 		name            string
 		tasks           PipelineTask
-		enableAPIFields bool
+		enableAPIFields string
```
Since this was bool earlier, the default was `false`. What is the default now? Is it an empty string? May be explicitly set `enableAPIFields` for all tests to the desired value instead of just setting to `beta` where needed?

JeromeJu replied:
Thanks that makes sense! I just change the condition to if not set just do not update the ctx configMap value, this might not require maintainers to edit every time even if it is just ctx.background. WDYT?

JeromeJu replied:
It seems that the reason for this change here is that resolver was an alpha field and this was previously used for setting alpha fields but now it has become beta. I think just in case that we are having more alpha features being used in this case, leaving it as `enableAPIFields` might leave more future room instead of changing back to enableAlphaFields or WithContext again?

lbernick commented on pkg/reconciler/pipelinerun/resources/pipelineref.go:
```diff
@@ -142,7 +142,6 @@ func readRuntimeObjectAsPipeline(ctx context.Context, namespace string, obj runt
 		// without actually creating the Pipeline on the cluster.
 		// Validation must happen before the v1beta1 Pipeline is converted into the storage version of the API,
 		// since validation of beta features differs between v1 and v1beta1
```
This comment is no longer relevant and can be removed. I think we can also move this function call out of the switch statement.

lbernick commented on pkg/apis/pipeline/v1beta1/taskref_validation.go:
```diff
@@ -35,6 +37,7 @@ func (ref *TaskRef) Validate(ctx context.Context) (errs *apis.FieldError) {
 	switch {
 	case ref.Resolver != "" || ref.Params != nil:
 		if ref.Resolver != "" {
+			errs = errs.Also(version.ValidateEnabledAPIFields(ctx, "resolver", config.BetaAPIFields).ViaField("resolver"))
```
can you please add tests for this change?

JeromeJu replied:
Thanks! Added and similarly for pipelineref test.

lbernick reviewed (commented):
I think this PR should also include the changes from https://github.com/tektoncd/pipeline/pull/6995, plus updating the docs on "beta features" and "alpha features" to clarify that they don't depend on api version.

JeromeJu commented on api_compatibility_policy.md:
```diff
@@ -90,19 +90,13 @@ For more information on support windows, see the [deprecations table](./docs/dep
 
+* `beta` (default) - This value indicates that only fields which are of `beta` (or greater) stability are enabled, i.e. `alpha` fields are not enabled.
 
-| Feature Versions -> | v1 | beta | alpha |
```
Thanks! I think the columns for the previous chart was indicating the apiVersion with `v1` and `beta` (indicating v1beta1). I agree that this is confusing and adding the first column to `enable-api-fields` value does make sense!

JeromeJu commented on pkg/apis/pipeline/v1/pipeline_validation.go:
```diff
@@ -55,7 +55,6 @@ func (p *Pipeline) Validate(ctx context.Context) *apis.FieldError {
 	// Validate beta fields when a Pipeline is defined, but not as part of validating a Pipeline spec.
 	// This prevents validation from failing when a Pipeline is converted to a different API version.
 	// See https://github.com/tektoncd/pipeline/issues/6616 for more information.
```
Thanks! This makes sense, I've tried to elaborate in https://github.com/tektoncd/pipeline/issues/7077 on why we need to move it into the member function of PipelineSpec. And I am planning to separate this move out as a 2nd commit in this Pull Request for the purpose of granularity.

lbernick commented on pkg/reconciler/pipelinerun/resources/pipelineref.go:
```diff
@@ -140,9 +140,6 @@ func readRuntimeObjectAsPipeline(ctx context.Context, namespace string, obj runt
 		vr := trustedresources.VerifyResource(ctx, obj, k8s, refSource, verificationPolicies)
 		// Issue a dry-run request to create the remote Pipeline, so that it can undergo validation from validating admission webhooks
 		// without actually creating the Pipeline on the cluster.
-		// Validation must happen before the v1beta1 Pipeline is converted into the storage version of the API,
```
The analogous comment should also be removed from taskref.go

lbernick commented:
please also update the chart in this docs section https://github.com/tektoncd/pipeline/blob/main/docs/additional-configs.md#beta-features

JeromeJu commented on pkg/apis/pipeline/v1beta1/task_validation.go:
```diff
@@ -65,7 +65,12 @@ func (t *Task) Validate(ctx context.Context) *apis.FieldError {
+	// Validate beta fields when a Pipeline is defined, but not as part of validating a Pipeline spec.
+	// This prevents validation from failing when a Pipeline is converted to a different API version.
+	// See https://github.com/tektoncd/pipeline/issues/6616 for more information.
+	errs = errs.Also(t.Spec.ValidateBetaFields(ctx))
```
Thanks. This was off during rebase.

JeromeJu commented on api_compatibility_policy.md:
```diff
@@ -90,26 +90,25 @@ For more information on support windows, see the [deprecations table](./docs/dep
 ## Feature Gates
 
-CRD API versions gate the overall stability of the CRD and its default behaviors. Within a particular CRD version, certain opt-in features may be at a lower stability level as described in [TEP-33](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md). These fields may be disabled by default and can be enabled by setting the right `enable-api-fields` feature-flag as d ...
+Stability levels of feature gates are independent from CRD apiVersions. The current group api-driven feature flag `enable-api-fields` gates the overall stability levels of features in the CRDs and their default behaviors. These fields may be disabled by default and can be enabled by setting the `enable-api-fields` feature-flag value:
```
Hmm for this part I was trying to call out api driven fields were guarded behind `enable-api-fields`. I think this would introduce ambiguity on whether other behavioural flags would also need to set `enable-api-fields` to `alpha` to get turned on? I wanted also to add some doc for behavioural flags after we introduce per feature flag but not to conflate them here. Please lemme know if that makes sense.

JeromeJu replied:
Thanks, updated!

JeromeJu commented on docs/additional-configs.md:
```diff
@@ -333,10 +333,10 @@ Features currently in "beta" are:
-| [Object Parameters and Results](pipelineruns.md#specifying-parameters) | [TEP-0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) | | [v0.46.0](https://github.com/tektoncd/pipeline/releases/tag/v0.46.0) | | No |
-| [Remote Tasks](./taskruns.md#remote-tasks) and [Remote Pipelines](./pipelineruns.md#remote-pipelines) | [TEP-0060](https://github.com/tektoncd/community/blob/main/teps/0060-remote-resolution.md) | | [v0.41.0](https://github.com/tektoncd/pipeline/releases/tag/v0.41.0) | | No |
-| [`Provenance` field in Status](pipeline-api.md#provenance)| [issue#5550](https://github.com/tektoncd/pipeline/issues/5550)| [v0.41.0](https://github.com/tektoncd/pipeline/releases/tag/v0.41.0)| [v0.48.0](https://github.com/tektoncd/pipeline/releases/tag/v0.48.0) | `enable-provenance-in-status`| No |
+| [Array Results and Array Indexing](pipelineruns.md#specifying-parameters)             | [TEP-0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md)     | [v0.38.0](https://github.com/tektoncd/pipeline/releases/tag/v0.38.0) | [v0.45.0](https://github.com/tektoncd/pipeline/releases/tag/v0.45.0) |                 | Yes |
```
Thanks!

JeromeJu commented:
/retest

JeromeJu commented:
/retest

JeromeJu commented:
When I rerun the pull-tekton-integration test:
[Previous 3 tests + 4 examples fails](https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/6941/pull-tekton-pipeline-integration-tests/1696915876025995264) -> [1 fails example tests](https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/6941/pull-tekton-pipeline-integration-tests/1696935435529359360)

JeromeJu commented:
/retest

JeromeJu commented:
Looks like there has been a job that succeeded from the CI job history.
/retest

JeromeJu commented:
/assign @pritidesai

JeromeJu commented:
/unhold 
[TEP0138](https://github.com/tektoncd/community/pull/1034) is merged.

Yongxuanzhang commented:
/assign

Yongxuanzhang commented on pkg/apis/pipeline/v1beta1/pipeline_validation.go:
```diff
@@ -91,6 +91,61 @@ func (ps *PipelineSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
+}
+
+// ValidateBetaFields returns an error if the Pipeline spec uses beta features but does not
+// have "enable-api-fields" set to "alpha" or "beta".
```
do you think it is ok to add a link to the list of our beta features here? So we can make sure the validation here covers all the features

JeromeJu replied:
Thanks for the suggestion @Yongxuanzhang  
I think the point for this func is to cover all existing beta features related with the pipelineSpec that are coupled before (the beta features that had different validations in v1 versus v1beta1). So not essentially all beta features here.

I understand the point here is to give users/cluster operators full notice of the change ahead with the transition to per-feature flag.
I've put a list in https://github.com/tektoncd/community/blob/main/teps/0138-decouple-api-and-feature-versioning.md as a cutoff of all existing beta/alpha features that are going to stay behind `enable-api-fields` and such validations. Does that sound good to add them in the api-compatibility policy update PR with Per-feature flag introduced to give heads up since this PR mainly handles the decoupling?

Yongxuanzhang replied:
Oh I think I misunderstood the code here, I thought it will cover all but the pipelineref is in another function 😂 I think it's fine.

Yongxuanzhang commented on pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:
```diff
@@ -3879,6 +3879,206 @@ func enableFeatures(t *testing.T, features []string) func(context.Context) conte
+			}},
+		},
+	}, {
+		name: "pipeline tasks - use of resolver without the feature flag set",
```
what does "without the feature flag set" mean here? We will set the enable-api-fields flag when running the cases

JeromeJu replied:
Oh thanks for the catch! This musth be left over with some refactoring. I corrected it and opened a minor edition to test names:
PTAL: https://github.com/tektoncd/pipeline/pull/7198
🙏

Yongxuanzhang commented on pkg/apis/pipeline/v1beta1/pipelineref_validation.go:
```diff
@@ -34,6 +35,7 @@ func (ref *PipelineRef) Validate(ctx context.Context) (errs *apis.FieldError) {
 
 	if ref.Resolver != "" || ref.Params != nil {
 		if ref.Resolver != "" {
+			errs = errs.Also(version.ValidateEnabledAPIFields(ctx, "resolver", config.BetaAPIFields).ViaField("resolver"))
```
can this be added to `ValidateBetaFields`?

JeromeJu replied:
I don't think that might be helpful for the codebase since that's pipelineRef rather than for the pipelineSpec?

Yongxuanzhang replied:
I see! Thanks!

JeromeJu commented:
👋  @tektoncd/core-maintainers  
It looks like this PR shall be ready to merge with some approvals already. But it would be much appreciated if there are more eyes 👀  on this from you since this is going to be a breaking change to some specific beta feature users according to our discussion in [TEP0138 ](https://github.com/tektoncd/community/blob/main/teps/0138-decouple-api-and-feature-versioning.md). 🙏  PTAL

Yongxuanzhang commented:
> 👋 @tektoncd/core-maintainers It looks like this PR shall be ready to merge with some approvals already. But it would be much appreciated if there are more eyes 👀 on this since this is going to be a breaking change to some specific beta feature users according to our discussion and consensus built in [TEP0138 ](https://github.com/tektoncd/community/blob/main/teps/0138-decouple-api-and-feature-versioning.md). 🙏
> 
> PTAL ❤️ 😸

Are we going to send emails to tekton users?

JeromeJu commented:
> > 👋 @tektoncd/core-maintainers It looks like this PR shall be ready to merge with some approvals already. But it would be much appreciated if there are more eyes 👀 on this since this is going to be a breaking change to some specific beta feature users according to our discussion and consensus built in [TEP0138 ](https://github.com/tektoncd/community/blob/main/teps/0138-decouple-api-and-feature-versioning.md). 🙏
> > PTAL ❤️ 😸
> 
> Are we going to send an email to tekton users?

Good call! We would need to send an email with the 0.53 LTS release clarifying the changes (decoupling feature&API versioning, introducing per-feature flags and the API compatibility policy updates) IIUC.

pritidesai commented:
Should we indicate the feature graduation process documented here only applies to the features driven by `enable-api-fields`? Does it apply to all existing and any new feature?

https://github.com/tektoncd/pipeline/blob/3819f4f48b326dee7ce644295a0a580d5674fd20/api_compatibility_policy.md#feature-graduation-process

I feel the graduation process is a mix of `enable-api-fields` and flag per feature.

> The dedicated feature flag for this feature will change the stability level for validation to beta.

This statement is from graduation process. Are we envisioning a flag for each new feature to have `alpha/beta/stable` values? If not, what are we going validate against?

JeromeJu commented:
Thanks for the review and the checking @pritidesai 

> I feel the graduation process is a mix of `enable-api-fields` and flag per feature.
> > The dedicated feature flag for this feature will change the stability level for validation to beta.

Thanks for pointing this out. I think `validation for` is misleading here. This won't apply to per-feature flag since we don't need to validate the stability of the feature being used by then. I've updated the wording here.

> Should we indicate the feature graduation process documented here only applies to the features driven by `enable-api-fields`? Does it apply to all existing and any new feature?
> I feel the graduation process is a mix of `enable-api-fields` and flag per feature.

The graduation process shall aim to be general, not to be specifically applied to API-driven fields or behavioural flags. It is some progress that all existing and new features in pipeline should go through IIUC. (like what we have as of today, a mix of behavioural flags and API-driven features in [beta features table in addtional-configs](https://github.com/tektoncd/pipeline/blob/main/docs/additional-configs.md#beta-features).

afrittoli commented:
/assign

JeromeJu commented:
/retest

pritidesai commented on pkg/apis/pipeline/v1/pipeline_validation.go:
```diff
@@ -91,6 +86,7 @@ func (ps *PipelineSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
 	errs = errs.Also(validateWhenExpressions(ps.Tasks, ps.Finally))
 	errs = errs.Also(validateMatrix(ctx, ps.Tasks).ViaField("tasks"))
 	errs = errs.Also(validateMatrix(ctx, ps.Finally).ViaField("finally"))
+	errs = errs.Also(ps.ValidateBetaFields(ctx))
```
NIT: should we consider this moving to the top as a first check? not that it stops any further validation but the error regarding the feature flag misconfiguration is up front.

JeromeJu replied:
Thanks for your time and the review! I've updated the PR according to the pointers 🙏

pritidesai commented on pkg/apis/pipeline/v1/task_validation.go:
```diff
@@ -98,6 +93,7 @@ func (ts *TaskSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
 	errs = errs.Also(validateTaskContextVariables(ctx, ts.Steps))
 	errs = errs.Also(validateTaskResultsVariables(ctx, ts.Steps, ts.Results))
 	errs = errs.Also(validateResults(ctx, ts.Results).ViaField("results"))
+	errs = errs.Also(ts.ValidateBetaFields(ctx))
```
NIT: similar concern - comment ⬆️

pritidesai commented on pkg/apis/pipeline/v1beta1/pipeline_validation.go:
```diff
@@ -89,6 +89,61 @@ func (ps *PipelineSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
+}
+
+// ValidateBetaFields returns an error if the Pipeline spec uses beta features but does not
+// have "enable-api-fields" set to "alpha" or "beta".
```
NIT: this comment is little confusing, suggest:

```
ValidateBetaFields returns an error if the PipelineSpec uses beta specifications governed by `enable-api-fields` but does not have "enable-api-fields" set to "alpha" or "beta".
```

pritidesai commented on pkg/apis/pipeline/v1beta1/pipeline_validation.go:
```diff
@@ -89,6 +89,61 @@ func (ps *PipelineSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
+		errs = errs.Also(pt.validateBetaFields(ctx).ViaFieldIndex("tasks", i))
+	}
+	for i, pt := range ps.Finally {
+		errs = errs.Also(pt.validateBetaFields(ctx).ViaFieldIndex("tasks", i))
```
s/tasks/finally/

afrittoli commented on pkg/apis/pipeline/v1/pipeline_validation.go:
```diff
@@ -58,17 +58,13 @@ func (p *Pipeline) Validate(ctx context.Context) *apis.FieldError {
 // Validate checks that taskNames in the Pipeline are valid and that the graph
 // of Tasks expressed in the Pipeline makes sense.
 func (ps *PipelineSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
+	errs = errs.Also(ps.ValidateBetaFields(ctx))
```
It sounds good - any special reason for that? 
We might have beta fields in the status as well, but those are produced by the controller, so I guess it's fine to skip validation.

JeromeJu replied:
Thanks for question Andrea. Related with the comment at https://github.com/tektoncd/pipeline/pull/6941#discussion_r1364619411 (many thanks to the suggestion from Priti), I think this was moved to the front for handling validations of easier logics first. And yes I do agree that validation for the PipelineSpec shall be sufficient for the webhook at the moment:)

afrittoli commented on api_compatibility_policy.md:
```diff
@@ -125,14 +124,12 @@ See the current list of [alpha features](https://github.com/tektoncd/pipeline/bl
+- GA/Stable features are enabled by default.
 
-- GA/Stable features are enabled by default
+- GA/Stable features are features that have been promoted from beta to the highest level of stability. They cannot be disabled in any CRD version.
```
These two sentences
> GA/Stable features are enabled by default.

and 

> GA/Stable features are features that have been promoted from beta to the highest level of stability. They cannot be disabled in any CRD version.

kind of contradict each other. "Enabled by default" implies that they can be disabled, but the default is "enabled".

I would say that stable features are not controlled by the `enable-api-fields` flag anymore.
I guess that in this context we are only talking about feature flags for API-driven features - other features may still be controlled by a separate dedicated feature flag even once they're stable.

afrittoli commented on pkg/apis/pipeline/v1/pipelinerun_validation.go:
```diff
@@ -66,11 +66,6 @@ func (ps *PipelineRunSpec) Validate(ctx context.Context) (errs *apis.FieldError)
-		// This prevents validation from failing in the reconciler when a Pipeline is converted to a different API version.
-		// See https://github.com/tektoncd/pipeline/issues/6616 for more information.
-		// TODO(#6592): Decouple API versioning from feature versioning
-		errs = errs.Also(ps.PipelineSpec.ValidateBetaFields(ctx).ViaField("pipelineSpec"))
```
Why is this removed?

JeromeJu replied:
Exactly, and I tried to include this in the commit message. Hope this captures it.

JeromeJu commented:
I've updated the PR and seems all current comments shall be resolved. Shall we merge this cc @afrittoli and @pritidesai ? 🙏

jerop commented:
/hold

want to make sure a non-googler takes a look (feel free to unhold)

cc @pritidesai @afrittoli

JeromeJu commented on pkg/apis/pipeline/v1beta1/pipeline_validation.go:
```diff
@@ -87,6 +92,60 @@ func (ps *PipelineSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
+		errs = errs.Also(pt.validateBetaFields(ctx).ViaFieldIndex("tasks", i))
+	}
+	for i, pt := range ps.Finally {
+		errs = errs.Also(pt.validateBetaFields(ctx).ViaFieldIndex("tasks", i))
```
I've updated it to finally, but as https://github.com/tektoncd/pipeline/issues/7258 tracks it actually does not work as expected. Will revisit when #7258 is fixed.

chitrangpatel replied:
Thanks!

chitrangpatel commented on pkg/apis/pipeline/v1beta1/pipeline_validation.go:
```diff
@@ -87,6 +92,60 @@ func (ps *PipelineSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
+			errs = errs.Also(version.ValidateEnabledAPIFields(ctx, "object results", config.BetaAPIFields).ViaFieldIndex("results", i))
+		case ResultsTypeArray:
+			errs = errs.Also(version.ValidateEnabledAPIFields(ctx, "array results", config.BetaAPIFields).ViaFieldIndex("results", i))
+		case ResultsTypeString:
```
probably a nit.

JeromeJu replied:
Yeah I also had similar question when this was added for v1 at https://github.com/tektoncd/pipeline/blob/8123d6332245008983aabe1dc910a63bd2c6f261/pkg/apis/pipeline/v1/pipeline_validation.go#L119 via https://github.com/tektoncd/pipeline/commit/1535587f5b45f4e228d1e2cb941379faef0fff20#diff-7ce49bf4f55e7af22bd0780d7f11c97c4fad98e7b6e0939fad1e81fc9e8e18fc. I think the reason that this was for more clarification of String Results are stable and other 2 types are beta features.

pritidesai commented on pkg/apis/pipeline/v1beta1/pipeline_validation.go:
```diff
@@ -92,6 +93,60 @@ func (ps *PipelineSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
+	}
+	for i, pt := range ps.Finally {
+		errs = errs.Also(pt.validateBetaFields(ctx).ViaFieldIndex("finally", i))
+	}
```
@JeromeJu I am sorry to go back on this but we had two new promotions - matrix and task level resources:

https://github.com/tektoncd/pipeline/blob/main/docs/additional-configs.md#beta-features

Do we need them here as well? or I think the features themselves have the check, is that sufficient?

JeromeJu replied:
I think this shall be sufficient. [Task-level resources beta promotion](https://github.com/tektoncd/pipeline/pull/7225/files#diff-da7c4e39bb1d27ceff9fb7f3fbe0c355525c2925bffc965ccb91adfc5389faf3) and [Matrix beta promotion](https://github.com/tektoncd/pipeline/pull/7164#discussion_r1342768960) shall already have the beta validations covered. 

And this function was meant to bring back the validations that were once missing at the stage TEP0138 was not in place and while feature & API versioning were coupled.

pritidesai commented on pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:
```diff
@@ -1557,3 +1557,205 @@ func TestPipelineRunWithTimeout_Validate(t *testing.T) {
 	}
 }
+
+func TestPipelineRunSpecBetaFeatures(t *testing.T) {
```
NIT: please add doc string explaining what is being tested

JeromeJu replied:
Thanks Priti. Added:)

pritidesai commented on pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:
```diff
@@ -4119,6 +4119,206 @@ func enableFeatures(t *testing.T, features []string) func(context.Context) conte
 	}
 }
 
+func TestPipelineWithBetaFields(t *testing.T) {
```
NIT: please add doc string what is being tested

pritidesai commented:
thanks @JeromeJu, the changes look good to me, I have left a few minor doc string comments in general and a question on recently promoted features.

JeromeJu commented:
> thanks @JeromeJu, the changes look good to me, I have left a few minor doc string comments in general and a question on recently promoted features.

Thanks @pritidesai . I've addressed the comments. Please lemme know if there are any other concerns? Or would you mind approving this given the release cutoff dates tmr 🙏

pritidesai commented:
thanks @JeromeJu for addressing all the comments 🙏 

/lgtm

pritidesai commented:
> /hold wait till [tektoncd/community#1034](https://github.com/tektoncd/community/pull/1034) gets merged first TBD on #6948

is there anything pending here? I think the required dependencies are addressed, please unhold this if there are no additional concerns.

JeromeJu commented:
Thanks for the time reviewing and the instructive comments ! @pritidesai 

/unhold

JeromeJu commented:
/retest
